internal/database/txn/transaction.go

```diff
@@ -170,7 +170,47 @@ func (t *RetryingTxnRunner) Txn(ctx context.Context, db *sqlair.DB, fn func(cont
 		}
 
 		if err := fn(ctx, tx); err != nil {
-			if rErr := t.retryStrategy(ctx, tx.Rollback); rErr != nil {
+			if rErr := t.retryStrategy(ctx, func(context.Context) error {
+				return tx.Rollback()
+			}); rErr != nil {
+				t.logger.Warningf("failed to rollback transaction: %v", rErr)
+			}
+			return errors.Trace(err)
+		}
+
+		return errors.Trace(t.commit(ctx, tx))
+	})
+}
+
+// TxnWithPrecheck runs a transaction with a precheck function that is executed
+// before the transaction is started. If the precheck function returns an error,
+// the transaction is not started. TxnWithPrecheck executes the input function
+// against the tracked database, using the sqlair package. The sqlair package
+// provides a mapping library for SQL queries and statements. Retry semantics
+// are applied automatically based on transient failures. This is the function
+// that almost all downstream database consumers should use.
+//
+// This should not be used directly, instead the TxnRunner should be used to
+// handle transactions.
+func (t *RetryingTxnRunner) TxnWithPrecheck(ctx context.Context, db *sqlair.DB, precheck func(context.Context) error, fn func(context.Context, *sqlair.TX) error) error {
+	return t.run(ctx, func(ctx context.Context) error {
+		// This is as close as we can get the precheck to the actual
+		// transaction. Anything closer, would interfere with the transaction
+		// itself.
+		if err := precheck(ctx); err != nil {
+			return errors.Trace(err)
```

Comment:
All failures here should always be a fatal failure.
```suggestion
			return errors.Join(errors.Trace(err), errAlwaysFatal)
```

Need to update `IsErrRetryable`

```go
const errAlwaysFatal errors.ConstError = "precheck failure is fatal"

func IsErrRetryable(err error) bool {
	if errors.Is(err, errAlwaysFatal) {
		return false
	}

	var dErr *driver.Error
	if errors.As(err, &dErr) && dErr.Code == driver.ErrBusy {
		return true
	}

	if errors.Is(err, sqlite3.ErrLocked) || errors.Is(err, sqlite3.ErrBusy) {
		return true
	}

	// Unwrap errors one at a time.
	for ; err != nil; err = errors.Unwrap(err) {
		errStr := err.Error()

		if strings.Contains(errStr, "database is locked") {
			return true
		}

		if strings.Contains(errStr, "cannot start a transaction within a transaction") {
			return true
		}

		if strings.Contains(errStr, "bad connection") {
			return true
		}

		if strings.Contains(errStr, "checkpoint in progress") {
			return true
		}
	}

	return false
}
```